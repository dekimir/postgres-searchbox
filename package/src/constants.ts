import type { Settings } from './index.types.js';
import type { ClientValidation } from './client.types.js';

export const VECTOR_COLUMN = 'postgres_searchbox_v1_doc';
export const INDEX_PREFIX = 'postgres_searchbox_v1_idx_';
export const MAX_REQ_PER_HTTP_REQ = 15;
export const MAX_HITS_PER_PAGE = 100;
export const MAX_PAGES = 100;
export const MAX_HITS_TOTAL = 3000;

const defaultSettings: Required<Settings> = {
  // * Attributes
  searchableAttributes: ['*'],
  attributesToRetrieve: ['*'],
  attributesForFaceting: ['*'],
  // * Faceting
  maxValuesPerFacet: 10,
  sortFacetValuesBy: 'count',
  // * Highlighting Snippeting
  attributesToHighlight: [],
  highlightPreTag: '__ais-highlight__',
  highlightPostTag: '__/ais-highlight__',
  // * Pagination
  hitsPerPage: 20,
  paginationLimitedTo: MAX_HITS_TOTAL,
  // * Performance
  numericAttributesForFiltering: ['*'],
  // * Advanced
  maxFacetHits: 100,
  renderingContent: {
    facetOrdering: {
      facets: { order: ['*'] },
    },
  },
};

const defaultClientValidation: Required<ClientValidation> = {
  // * Attributes
  validAttributesToRetrieve: ['*'],
  // * Filtering
  validFacetFilters: ['*'],
  // * Highlighting Snippeting
  validAttributesToHighlight: ['*'],
  validHighlightPreTags: ['__ais-highlight__'],
  validHighlightPostTags: ['__/ais-highlight__'],
  // * Pagination
  maxHitsPerPage: MAX_HITS_PER_PAGE,
  maxPage: MAX_PAGES,
  maxOffset: MAX_HITS_TOTAL,
  maxLength: MAX_HITS_PER_PAGE,
  maxHitsTotal: MAX_HITS_TOTAL,
};

export const defaults = {
  settings: defaultSettings,
  clientValidation: defaultClientValidation,
};
