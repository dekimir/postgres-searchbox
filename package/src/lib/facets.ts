import format from 'pg-format';
import type { Props, GetFacetsReturn } from './facets.types.js';

/**
 * Ambiguous params
 * maxValuesPerFacet: The maximum number of facet values to return for each facet in a regular search.
 * sortFacetValuesBy: Controls how facet values are fetched.
 * maxFacetHits: Maximum number of facet hits to return during a search for facet values.
 */

export const getFacets = ({
  facets: facetsMaybeString,
  attributesForFaceting,
  maxValuesPerFacet,
  sortFacetValuesBy,
  maxFacetHits,
  renderingContent,
  numericAttributesForFiltering,
}: Props): GetFacetsReturn => {
  // Build facets based on 2 conditionals
  const facets: readonly string[] =
    facetsMaybeString?.includes('*') || !facetsMaybeString
      ? attributesForFaceting // facetsMaybeString is * or undefined
      : Array.isArray(facetsMaybeString) // is it an array already?
      ? facetsMaybeString // yes
      : [facetsMaybeString]; // no, string cast to array

  if (!facets?.length) return null;

  const limit = Number.isInteger(maxValuesPerFacet) ? maxValuesPerFacet : 10;

  // For things like
  const cte = facets
    .map((facet) =>
      format(
        /* sql */ `
        %I AS (
          SELECT coalesce( 
            json_object_agg( %s, postgres_searchbox_v1_count ORDER BY postgres_searchbox_v1_count DESC ), 
            '{}'::json
          ) as details
            FROM (
              SELECT %s, count(*) AS postgres_searchbox_v1_count
              FROM all_selection
              WHERE %s IS NOT NULL
              GROUP by %s
              ORDER BY ${
                sortFacetValuesBy === 'count' ? `count(*) DESC` : `%s ASC`
              } 
              LIMIT ${limit}
            ) t
        )
      `,
        `${facet}_selection`,
        facet,
        facet,
        facet,
        facet,
        facet
      )
    )
    .join(',');

  const statsCte = numericAttributesForFiltering
    ?.filter((f) => f !== '*')
    .map((facet) =>
      format(
        /* sql */ `
        %I AS (
          SELECT 
              json_build_object( 
                'max', MAX (%I::numeric), 
                'min', MIN(%I::numeric), 
                'avg', round( AVG(%I::numeric), 4), 
                'sum', SUM(%I::numeric)
              ) 
          FROM all_selection
          WHERE %I IS NOT NULL 
        )
      `,
        `${facet}_stats`,
        facet,
        facet,
        facet,
        facet,
        facet
      )
    )
    .join(',');

  const json = /* sql */ `
    'facets', json_build_object(
      ${facets
        .map((facet) =>
          format(`%L, (SELECT details FROM %I)`, facet, `${facet}_selection`)
        )
        .join(',')}
    )`;

  const statsJson = /* sql */ `
    'facets_stats', json_build_object(
      ${numericAttributesForFiltering
        ?.filter((f) => f !== '*')
        .map((facet) =>
          format(`%L, (SELECT * FROM %I)`, facet, `${facet}_stats`)
        )
        .join(',')}
    )`;

  const order = renderingContent?.facetOrdering?.facets?.order;
  const values = renderingContent?.facetOrdering?.values;

  // Build up the object based on the facets
  let newOrder: string[] | undefined;
  if (order?.includes('*')) {
    newOrder = facets.map((f) => f);
  }

  return {
    db: { cte, statsCte, json, statsJson },
    renderingContent: {
      ...renderingContent,
      facetOrdering: {
        ...renderingContent?.facetOrdering,
        facets: {
          order: newOrder || order,
        },
        // TODO values
        values,
      },
    },
  };
};
