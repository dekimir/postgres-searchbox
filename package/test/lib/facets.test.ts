import pkg from 'pg';
const { Client } = pkg;
import format from 'pg-format';
// Scripts
import { initTestDatabase } from '@scripts/mock-data.js';
import { createColumnAndIndex } from '@scripts/create-index.js';
// Lib
import { getFacets } from '@/lib/facets.js';
// Constants
import { defaults } from '@/constants.js';

/**
 * Test facets
 */

describe('facets', () => {
  const tableName = 'test_table';
  const initTestDatabaseParams = { tableName, rowCount: 100, fakerSeed: 123 };

  const client = new Client();
  client.connect();

  beforeEach(async () => {
    // Not all tests use the database so don't drop the table here
    // like in other test files
  });

  afterAll(async () => {
    const dropSql = format('DROP TABLE IF EXISTS %I', tableName);
    await client.query(dropSql);
    await client.end();
  });

  it('should return workingSQL fragments', async () => {
    // Drop the table so that the tests can be run independently
    const dropSql = format(`DROP TABLE IF EXISTS %I`, tableName);
    await client.query(dropSql);
    // Prerequisites
    await initTestDatabase(initTestDatabaseParams);
    await createColumnAndIndex({ tableName });

    const facets = getFacets({
      ...defaults.settings,
      facets: ['brand', 'price'],
      numericAttributesForFiltering: ['price'],
    });

    const { cte, statsCte, json, statsJson } = facets?.db || {};

    expect(typeof cte).toBe('string');
    expect(typeof statsCte).toBe('string');
    expect(typeof json).toBe('string');
    expect(typeof statsJson).toBe('string');

    /**
     * Minimal example of the main SQL query in index.ts
     */

    const formattedSql = format(
      /* sql */ `(
      -- Step 1: Get all results
      WITH all_selection AS (
        SELECT * FROM %I
      ),
      hits_selection AS (
        SELECT * FROM all_selection LIMIT 10
      ),
      -- Step 2: Get the counts for each facet
      ${cte},
      -- Step 3: Get facets_stats on numeric attributes
      ${statsCte}
      -- Step 4: Return it all as a JSON object
      SELECT json_build_object(
        'totalHits', ( SELECT count(*) FROM all_selection ),
        'hits', jsonb_agg(hits_selection.*)::json,
        ${json},
        ${statsJson}
      ) AS "json"
      FROM hits_selection
    )`,
      tableName
    );

    // console.log(formattedSql);

    const result = await client.query(formattedSql);

    expect(result.rows).toHaveLength(1);
    expect(result.rows[0].json.totalHits).toBe(100);
    expect(result.rows[0].json.hits).toHaveLength(10);
    expect(result.rows[0].json.facets.brand).toMatchSnapshot();
    expect(result.rows[0].json.facets.price).toMatchSnapshot();
    expect(result.rows[0].json.facets_stats.price).toMatchSnapshot();
  });
});
