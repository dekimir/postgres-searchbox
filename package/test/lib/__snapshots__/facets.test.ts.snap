// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`facets should return workingSQL fragments 1`] = `
{
  "Altenwerth LLC": 2,
  "Bauch - Leannon": 2,
  "Beier Group": 2,
  "Dibbert - O'Kon": 2,
  "Dibbert LLC": 2,
  "Donnelly - Klocko": 2,
  "Emmerich and Sons": 2,
  "Schuster Group": 2,
  "Sipes, O'Connell and Dooley": 2,
  "Stracke and Sons": 2,
}
`;

exports[`facets should return workingSQL fragments 2`] = `
{
  "11": 6,
  "15": 5,
  "18": 5,
  "20": 4,
  "22": 4,
  "23": 5,
  "27": 7,
  "30": 5,
  "31": 4,
  "40": 5,
}
`;

exports[`facets should return workingSQL fragments 3`] = `
{
  "avg": 24.26,
  "max": 40,
  "min": 10,
  "sum": 2426,
}
`;
